src/picongpu/include/particles/ionization/byCollision/ThomasFermi/AlgorithmThomasFermi.hpp

```diff
@@ -0,0 +1,158 @@
+        HDINLINE void
+        operator()( const KinEnergyDensityType kinEnergyDensity, const DensityType density, ParticleType& parentIon, float_X randNr )
+        {
+
+            /* @TODO replace the float_64 with float_X and make sure the values are scaled to PIConGPU units */
+            const float_64 protonNumber = GetAtomicNumbers<ParticleType>::type::numberOfProtons;
+            const float_64 neutronNumber = GetAtomicNumbers<ParticleType>::type::numberOfNeutrons;
+            float_64 chargeState = attribute::getChargeState(parentIon);
+
+            /* ionization condition */
+            if (chargeState < protonNumber)
+            {
+                /* atomic mass number (usually A) A = N + Z */
+                float_64 massNumber = neutronNumber + protonNumber;
+
+                /** @TODO replace the static_cast<float_64> by casts to float_X
+                 * or leave out entirely and compute everything in PIConGPU scaled units
+                 */
+                float_64 const densityUnit = static_cast<float_64>(particleToGrid::derivedAttributes::Density().getUnit()[0]);
+                float_64 const kinEnergyDensityUnit = static_cast<float_64>(particleToGrid::derivedAttributes::EnergyDensity().getUnit()[0]);
+                /* convert from kinetic energy density to average kinetic energy per particle */
+                float_64 kinEnergy = (kinEnergyDensity / density) * (kinEnergyDensityUnit / densityUnit);
+                /** convert kinetic energy in J to "temperature" in eV by assuming an ideal electron gas
+                 * E_kin = 3/2 k*T
+                 */
+                float_64 temperature = kinEnergy * UNITCONV_Joule_to_keV * float_64(1.e3) * float_64(2./3.);
+
+                float_64 T_0 = temperature/math::pow(protonNumber,float_64(4./3.));
+
+                float_64 T_F = T_0 / (float_64(1.) + T_0);
+
+                /* for all the fitting parameters @see TFFittingParameters.def */
+
+                /** this is weird - I have to define temporary variables because
+                 * otherwise the math::pow function won't recognize those at the
+                 * exponent position */
+                float_64 TFA2_temp = TFA2;
+                float_64 TFA4_temp = TFA4;
+                float_64 TFBeta_temp = TFBeta;
+
+                float_64 A = TFA1 * math::pow(T_0,TFA2_temp) + TFA3 * math::pow(T_0,TFA4_temp);
+
+                float_64 B = -math::exp(TFB0 + TFB1*T_F + TFB2*math::pow(T_F,float_64(7.)));
+
+                float_64 C = TFC1 * T_F + TFC2;
+
+                /** requires mass density in g/cm^3
+                 * @TODO relocate constants to param file or leave out and calculate unitless
+                 */
+                float_64 const nAvogadro = 6.022e23;
+                float_64 const convM3ToCM3 = 1.e6;
+
+                float_64 massDensity = density * densityUnit * massNumber / nAvogadro / convM3ToCM3;
```

Comment:
besides `density * densityUnit` you can use an other `constexpr` here to reduce RT calculations